spec.bs

```diff
@@ -6,57 +6,882 @@ Status: CG-DRAFT
+        {url: "blob:https://c.example/789…"}
+      ],
+      { data: { name: 'experimentA' } }
+    );
+
+    // Assumes that the fenced frame 'my-fenced-frame' has already been attached.
+    document.getElementById('my-fenced-frame').config = fencedFrameConfig;
+  </pre>
+
+  inside the `experiment.js` worklet script:
+  <pre class="lang-js">
+    class SelectURLOperation {
+      hash(experimentName, seed) { … }
+
+      async run(data, urls) {
+        const seed = await this.sharedStorage.get('seed');
+        return hash(data.name, seed) % urls.length;
+      }
+    }
+    register('select-url-for-experiment', SelectURLOperation);
+  </pre>
+</div>
+
+The {{SharedStorageWorklet}} Interface {#worklet}
+=================================================
+The {{SharedStorageWorklet}} object allows developers to supply [=module scripts=] to process [=Shared Storage=] data and then output the result through one or more of the output gates. Currently there are two output gates, the [=private aggregation=] output gate and the {{WindowSharedStorage/selectURL()|URL-selection}} output gate.
 
-Shared Storage DOM Interface {#dom-interface}
-=============================================
 <xmp class='idl'>
+  [Exposed=(Window)]
+  interface SharedStorageWorklet : Worklet {
+  };
+</xmp>
+
+  ## The {{SharedStorageWorkletGlobalScope}} ## {#global-scope}
+
+  The {{SharedStorageWorklet}}'s [=worklet global scope type=] is {{SharedStorageWorkletGlobalScope}}. 
+
+  <xmp class='idl'>
+    [Exposed=SharedStorageWorklet, Global=SharedStorageWorklet]
+    interface SharedStorageWorkletGlobalScope : WorkletGlobalScope {
+      undefined register(DOMString name,
+                         SharedStorageOperationConstructor operationCtor);
+    };
+
+    callback SharedStorageOperationConstructor = SharedStorageOperation(object options);
+
+    [Exposed=SharedStorageWorklet]
+    interface SharedStorageOperation {
+    };
+  </xmp>
+
+  Each {{SharedStorageWorkletGlobalScope}} has an associated [=environment settings object=] <dfn for=SharedStorageWorkletGlobalScope>outside settings</dfn>, which is the associated {{SharedStorageWorklet}}'s [=relevant settings object=].
+
+  The {{SharedStorageWorkletGlobalScope}}'s [=module map=]'s [=module scripts=] should each define and {{register}} one or more {{SharedStorageOperation}}s. 
+
+  Each {{SharedStorageWorkletGlobalScope}} also has an associated <dfn for=SharedStorageWorkletGlobalScope>operation map</dfn>, which is an [=map=], initially empty, of [=strings=] (denoting operation names) to [=functions=].
+
+  Currently each {{SharedStorageOperation}} registered via {{SharedStorageWorkletGlobalScope/register()}} must be one of the following two types:
+  * {{SharedStorageRunOperation}}
+  * {{SharedStorageSelectURLOperation}}
+
+  A {{SharedStorageRunOperation}} is an {{SharedStorageOperation}} that has no JavaScript return value, and hence is especially suited for calling into the [=private aggregation=] service.
```

Comment:
What does it mean to have "no JavaScript return value"?  Maybe it would be better to say the return value is ignored and most of these operations are expected to return undefined.

I might also phrase the second half of this sentence as a separate sentence:

"The {{SharedStorageRunOperation}} is designed to work with output gates that do not need a return value, like the [=private aggregation=] service."

Reply:
Hmm, it does seem that the returned promise might be relevant as a way to signal completion of the operation.  Its just the fulfilled promise value is ignored.

Reply:
Rephrased.